### PR TITLE
Custom word boundary

### PR DESCRIPTION
**Pull request by Mikk3lRo**
### Summary

The default / current word boundary regex `^|\\b|\\s` is remarkably bad at actually finding word boundaries in languages that use non-unicode characters. A word boundary is basically detected after any non-ascii character (fx. ü, å, ø and æ to mention just a few - but there are MANY).

I've looked into possibilities, and unfortunately there doesn't seem to be any way to get decent word-boundary detection for anything except ascii in javascripts RegExp implementation... without either using a third-party library or including some 4k+ characters in the string.

Therefore, I don't see any way to reliably detect word boundaries with any pre-set, hardcoded regex.

Turning it into an option means that people can at least set something appropriate for their individual language and / or use case if they care about word boundaries being detected in "weird" places.

Please double-check that:

  - [x] All changes were made in CoffeeScript files, **not** JavaScript files.
  - [x] You used [Grunt](https://github.com/harvesthq/chosen/blob/master/contributing.md#grunt) to build the JavaScript files and tested them locally.
  - [x] You've updated both the jQuery *and* Prototype versions.
  - [x] You haven't manually updated the version number in `package.json`.
  - [x] If necessary, you've updated [the documentation](https://github.com/harvesthq/chosen/blob/master/public/options.html).

### References

First partial PR from: https://github.com/harvesthq/chosen/pull/2894

### CODE REVIEW

**Comment by Mikk3lRo** [comment]
I'm fairly certain this will solve at least this open issue / bug:
https://github.com/harvesthq/chosen/issues/2862

**Comment by tjschuck** [comment]
@Mikk3lRo Can you squash this down to one commit?  [Instructions here](http://blog.steveklabnik.com/posts/2012-11-08-how-to-squash-commits-in-a-github-pull-request).

File: coffee/lib/abstract-chosen.coffee [hunk 1]
```diff
@@ -27,6 +27,7 @@ class AbstractChosen
     @enable_split_word_search = if @options.enable_split_word_search? then @options.enable_split_word_search else true
     @group_search = if @options.group_search? then @options.group_search else true
     @search_contains = @options.search_contains || false
+    @search_word_boundary = if @options.search_word_boundary? then @options.search_word_boundary else '^|\\s|\\b'
```
Comment (koenpunt) [inline coffee/lib/abstract-chosen.coffee]: Can’t this be just;

```coffee
@search_word_boundary = options.search_word_boundary || '^|\\s|\\b'
``` 

?

Comment (koenpunt) [reply coffee/lib/abstract-chosen.coffee]: Also, the option should not necessarily include “search” in its name

File: public/options.html [hunk 2]
```diff
@@ -87,11 +87,21 @@
...
           <td>group_search</td>
           <td>true</td>
           <td>By default, Chosen will search group labels as well as options, and filter to show all options below matching groups. Set this to <code class="language-javascript">false</code> to search only in the options.</td>
         </tr>
         <tr>
+          <td>search_word_boundary</td>
+          <td>^|\\b|\\s</td>
+          <td>By default, Chosen uses JS RegExp's built-in word boundary to detect word beginnings as well as whitespace or the beginning of the entire label. That works great for ascii-only languages, but <strong>will</strong> erroneously detect word boundaries after letters with umlauts among many, many others.<br>You can pass a string (that will be interpreted as part of a <code class="language-javascript">RegExp</code>) refined for your language and use case to correctly detect word boundaries. A (simplified) example could be <code class="language-javascript">'^|[^A-zæøåÆØÅ]'</code> for Danish.</td>
+        </tr>
+        <tr>
```
Comment (koenpunt) [inline public/options.html]: This is now in here twice

File: spec/jquery/searching.spec.coffee [hunk 3]
```diff
@@ -247,7 +247,7 @@ describe "Searching", ->
 
     expect(div.find(".active-result").length).toBe(1)
     expect(div.find(".active-result")[0].innerHTML).toBe("oh <em>h</em>ello")
-
+
```
Comment (koenpunt) [inline spec/jquery/searching.spec.coffee]: Nit; unrelated change

Comment (Mikk3lRo) [reply spec/jquery/searching.spec.coffee]: Yeah... not eactly a git wiz here, and the squahing turnied out to not be as easy as the instuctions suggested due to a conflict on the options page. I'm just about to give up and start a new branch instead.

File: coffee/lib/abstract-chosen.coffee (hunk 1 above)
Comment (Mikk3lRo) [reply coffee/lib/abstract-chosen.coffee]: Is it not logical to prefix search related options with "search"?

And I'm pretty sure the question mark is needed, otherwise the empty string will evaluate to false and the default value will be used. The empty string is a perfectly valid value for this option... if set to an empty string Chosen works like `search_contains=true`, but allowing `enable_split_word_search` to decide whether to only match the beginning of the option - see linked issue for a relevant use case.

File: spec/jquery/searching.spec.coffee (hunk 3 above)
Comment (Mikk3lRo) [reply spec/jquery/searching.spec.coffee]: Ok the whole thing is a mess now. I have no idea how to recover. Can I start a new branch and pull request?

**Comment by Mikk3lRo** [comment]
Sorry, but this is too messed up now. Some stupid conflict going around itself in circles.
